lib/src/signed_video_h26x_sign.c

```diff
@@ -524,32 +525,48 @@ signed_video_add_nalu_part_for_signing_with_timestamp(signed_video_t *self,
       self->frame_count++;  // It is ok for this variable to wrap around
     }
 
-    SV_THROW(hash_and_add(self, &nalu));
-    // Depending on the input NALU, we need to take different actions. If the input is an I-NALU we
-    // have a transition to a new GOP. Then we need to generate the necessary SEI-NALU(s) and put in
-    // prepend_list. For all other valid NALUs, simply hash and proceed.
-    if (nalu.is_first_nalu_in_gop && nalu.is_last_nalu_part) {
-      // An I-NALU indicates the start of a new GOP, hence prepend with SEI-NALUs. This also means
-      // that the signing feature is present.
-
-      // Store the timestamp for the first nalu in gop.
-      if (timestamp) {
-        self->gop_info->timestamp = *timestamp;
-        self->gop_info->has_timestamp = true;
-      }
-
-      uint8_t *payload = NULL;
-      uint8_t *payload_signature_ptr = NULL;
-
-      SV_THROW(generate_sei_nalu(self, &payload, &payload_signature_ptr));
-      // Add |payload| to buffer. Will be picked up again when the signature has been generated.
-      add_payload_to_buffer(self, payload, payload_signature_ptr);
-      // Now we are done with the previous GOP. The gop_hash was reset right after signing and
-      // adding it to the SEI NALU. Now it is time to start a new GOP, that is, hash and add this
-      // first NALU of the GOP.
+    // Process the NALU based on the presence of linked hash and whether it is the first NALU in the
+    // GOP
+    if (!(self->gop_info->has_linked_hash && nalu.is_first_nalu_in_gop)) {
```

Comment:
This flag stuff is really difficult to follow. As a first step I suggest to split the code into one if as
```
if (!self->linked_hash_on) {
  // The code as it is present today from "hash_and_add to hash_and_add"
} else {
  // Add the "new" code with your flag technique
}
```
This will make it easier to read, but most important the code becomes more robust to bugs